src/openfermion/utils/_qubit_tapering_from_stabilizer.py

```diff
@@ -0,0 +1,406 @@
+    """
+    Perform the term reduction using stabilizer conditions.
+
+    Auxiliary function to reduce_number_of_terms that returns the
+    Pauli strings with the same length as in the starting operator.
+    """
+    term_list_duplicate = list(terms.terms)
+    term_list = [QubitOperator(x) for x in term_list_duplicate]
+
+    if manual_input is False:
+        fixed_positions = []
+
+    for i, x in enumerate(stabilizer_list):
+        selected_stab = list(stabilizer_list[0].terms)[0]
+
+        if manual_input is False:
+            # Finds qubit position and its Pauli.
+            for qubit_pauli in selected_stab:
+                if qubit_pauli[0] not in fixed_positions:
+                    fixed_positions += [qubit_pauli[0]]
+                    fixed_op = qubit_pauli[1]
+                    break
+        else:
+            # Finds Pauli of the fixed qubit.
+            for qubit_pauli in selected_stab:
+                if qubit_pauli[0] == fixed_positions[i]:
+                    fixed_op = qubit_pauli[1]
+                    break
+
+        if fixed_op in ['X', 'Z']:
+            other_op = 'Y'
+        else:
+            other_op = 'X'
+
+        new_list = []
+        updated_stabilizers = []
+        for y in term_list:
+            new_list += [_fix_single_term(y, fixed_positions[i], fixed_op,
+                                          other_op, stabilizer_list[0])]
+        for update_stab in stabilizer_list[1:]:
+            updated_stabilizers += [_fix_single_term(update_stab,
+                                                     fixed_positions[i],
+                                                     fixed_op, other_op,
+                                                     stabilizer_list[0])]
+        term_list = new_list
+        stabilizer_list = updated_stabilizers
+
+        _check_stabilizer_linearity(stabilizer_list,
+                                    msg='Linearly dependent stabilizers.')
+        _check_commuting_stabilizers(stabilizer_list,
+                                     msg='Stabilizers anti-commute.')
+
+    new_terms = QubitOperator()
+    for x, ent in enumerate(term_list):
+        new_terms += ent * terms.terms[term_list_duplicate[x]]
+    for x, ent in enumerate(term_list):
+        term_list_duplicate[x] = (
+            QubitOperator(term_list_duplicate[x]) /
+            list(ent.terms.items())[0][1])
+        term_list[x] = list(ent.terms)[0]
+
+    even_newer_terms = QubitOperator()
+    for pauli_string, coefficient in new_terms.terms.items():
+        even_newer_terms += coefficient * _lookup_term(
+            pauli_string,
+            term_list_duplicate,
+            term_list)
+
+    return even_newer_terms, fixed_positions
+
+
+def reduce_number_of_terms(operator, stabilizers,
+                           maintain_length=False,
+                           output_fixed_positions=False,
+                           manual_input=False,
+                           fixed_positions=[]):

```

Comment:
It is unlikely that this method will be referenced abstractly later but for safety you should remove the list optional type in exchange for something that won't be mutable.